manuals/mm8d-hwtest.py.1.in
.TH "mm8d-hwtest.py" "1" "@version@" "Pozsár Zsolt" "MM8D"
.SH "NAME"
.LP 
mm8d-hwtest.py - hardware test program
.SH "SYNTAX"
.LP
\fBmm8d-hwtest.py\fP
.SH "DESCRIPTION"
.LP 
This is a part of MM8D growing house and irrigation controlling and remote monitoring
system.
.PP
This is a menu-controlled test program for check inputs, relay and open collector
outputs of the interface board via GPIO or LPT port and you can write text to the
mini serial console with it.
Before use, run \fBmm8d-stopdaemon\fP script to stop daemon and after test
run \fBmm8d-startdaemon\fP to normal operating.
.SH "FILES"
.LP
.TP
\fI/etc/mm8d/mm8d.ini\fR or \fI/usr/local/etc/mm8d/mm8d.ini\fR
main configuration file
.SH "EXIT STATUS"
.TP
.B 0
normal exit
.TP
.B 1
cannot open configuration file
.TP
.B 17
cannot access I/O port
.SH "HOMEPAGE"
\fIhttp://www.pozsarzs.hu\fR
.SH "SEE ALSO"
.PD 0
.LP
\fBgetcameras.cgi\fP(7)
\fBgetchannel.cgi\fP(7)
\fBgetdata.cgi\fP(7)
\fBgetenvirconf.cgi\fP(7)
\fBgetsupply.cgi\fP(7)
\fBmm8d-creatediagrams\fP(1)
\fBmm8d-editenvirconf\fP(1)
\fBmm8d-editmainconf\fP(1)
\fBmm8d-exportlog2csv\fP(1)
\fBmm8d-getsnapshots-seccam\fP(1)
\fBmm8d-getsnapshots-tentcam\fP(1)
\fBmm8d-maintainlog\fP(1)
\fBmm8d-msctest.py\fP(1)
\fBmm8d-override\fP(1)
\fBmm8d.py\fP(8)
\fBmm8d-restartdaemon\fP(1)
\fBmm8d-startdaemon\fP(1)
\fBmm8d-statusofdaemon\fP(1)
\fBmm8d-stopdaemon\fP(1)
\fBmm8d-updatestartpage\fP(1)
\fBmm8d-viewlog\fP(1)
\fBmm8d-webpage\fP(1)
and \fI/usr/share/doc/mm8d/\fP or equivalent on your system.
.SH "AUTHOR"
Pozsár Zsolt \fIpozsarzs@gmail.com\fR.


documents/changelog
 MM8D (0.5) unstable; urgency=low

  * Some bug fixes.
  * Electricity consumption meter support (MM10D).
  * Water consumption meter support (MM11D).
  * Improved web interface.
  * Modified main configuration file and configuration programs.
  * Modified log file reader.
  * Export log file to CSV.
  * Take and display security camera snapshots on the web interface (this is not the marquee camera on the channel side).
  * Prepared ModBUS/TCP support.
  * Cannot exits without active channels.

 -- Pozsár Zsolt <pozsarzs@gmail.com>             XXX, 00 May 2023 00:00:00

 MM8D (0.4) unstable; urgency=low

  * Add mini serial console support. See miniserialconsole.txt for
    details.

 -- Pozsár Zsolt <pozsarzs@gmail.com>             Tue, 13 Dec 2022 14:05:12

 MM8D (0.3.1) unstable; urgency=low

  * Some bugfix.

 -- Pozsár Zsolt <pozsarzs@gmail.com>             Fri, 19 Aug 2022 15:44:00

 MM8D (0.3) unstable; urgency=low

  * New option is in envir-ch?.ini files.
  * Irrigator control is included.
  * Some bugfix.
  * Buttons for get environment characteristic of growing houses.
  * Modified function of inputs (on interface board).
  * New function of outputs (on interface board).

 -- Pozsár Zsolt <pozsarzs@gmail.com>             Mon, 28 Mar 2022 16:57:00

 MM8D (0.2) unstable; urgency=low

  * New section (Override outputs) on webpage.
  * New output format (XML) of getdata.cgi.
  * Two new utility scripts with manual page.
  * Packages for amd64 architecture.
  * Some bugfix.

 -- Pozsár Zsolt <pozsarzs@gmail.com>             Fri, 31 Dec 2021 23:59:59


 MM8D (0.1) unstable; urgency=low

  * Initial release.

 -- Pozsár Zsolt <pozsarzs@gmail.com>             Thu, 10 Jun 2021 08:28:00


programs/mm8d-hwtest.py
#!/usr/bin/python3
# +----------------------------------------------------------------------------+
# | MM8D v0.5 * Growing house and irrigation controlling and monitoring system |
# | Copyright (C) 2020-2023 Pozsar Zsolt <pozsarzs@gmail.com>                  |
# | mm8d-hwtest.py                                                             |
# | Hardware test program                                                      |
# +----------------------------------------------------------------------------+

#   This program is free software: you can redistribute it and/or modify it
# under the terms of the European Union Public License 1.2 version.
#
#   This program is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE.

# Exit codes:
#   0: normal exit
#   1: cannot open configuration file
#  17: cannot access i/o port

import configparser
import io
import os
import platform
import serial
import sys
import time

# constants
arch=platform.machine()
if arch.find("86") > -1:
  HW = 1
  import portio
else:
  HW = 0
  import RPi.GPIO as GPIO

USRLOCALDIR = 1
if (USRLOCALDIR == 1):
  CONFFILE = '/usr/local/etc/mm8d/mm8d.ini'
else:
  CONFFILE = '/etc/mm8d/mm8d.ini'

global EOL
global LPTADDRESSES
EOL = "\r"
LPTADDRESSES = [0x378,0x278,0x3bc]

# load configuration
def loadconfiguration(CONFFILE):
  global com
  global prt_com
  global com_speed
  C = 'COMport'
  G = 'GPIOport'
  L = 'LPTport'
  if HW == 0:
    global prt_i1
    global prt_i2
    global prt_i3
    global prt_i4
    global prt_ro1
    global prt_ro2
    global prt_ro3
    global prt_ro4
    global prt_lo1
    global prt_lo2
    global prt_lo3
    global prt_lo4
  else:
    global prt_lpt
  try:
    with open(CONFFILE) as f:
      mm8d_config = f.read()
    config = configparser.RawConfigParser(allow_no_value=True)
    config.read_file(io.StringIO(mm8d_config))
    # number of the used GPIO ports
    if HW == 0:
      prt_i1 = int(config.get(G,'prt_i1'))
      prt_i2 = int(config.get(G,'prt_i2'))
      prt_i3 = int(config.get(G,'prt_i3'))
      prt_i4 = int(config.get(G,'prt_i4'))
      prt_ro1 = int(config.get(G,'prt_ro1'))
      prt_ro2 = int(config.get(G,'prt_ro2'))
      prt_ro3 = int(config.get(G,'prt_ro3'))
      prt_ro4 = int(config.get(G,'prt_ro4'))
      prt_lo1 = int(config.get(G,'prt_lo1'))
      prt_lo2 = int(config.get(G,'prt_lo2'))
      prt_lo3 = int(config.get(G,'prt_lo3'))
      prt_lo4 = int(config.get(G,'prt_lo4'))
    else:
      # address of the used LPT port (0x378: 0, 0x278: 1, 0x3BC: 2)
      prt_lpt = int(config.get(L,'prt_lpt'))
      if (prt_lpt < 0) or (prt_lpt > 2):
        prt_lpt = 0
    # port name
    prt_com = config.get(C,'prt_com')
    # port speed
    com_speed = 9600
    com_speed = int(config.get(C,'com_speed'))
  except:
    print("ERROR #1: Cannot open configuration file!");
    sys.exit(1);

# main function
print("\nMM8D hardware test utility * (C) 2020-2023 Pozsar Zsolt")
print("-------------------------------------------------------")
if os.getuid():
  print(" * You need to be root!")
  sys.exit(0)
print(" * load configuration: %s..." % CONFFILE)
loadconfiguration(CONFFILE)
print(" * setting ports...")
com = serial.Serial(prt_com, com_speed)
if HW == 0:
  GPIO.setwarnings(False)
  GPIO.setmode(GPIO.BCM)
  GPIO.setup(prt_i1,GPIO.IN)
  GPIO.setup(prt_i2,GPIO.IN)
  GPIO.setup(prt_i3,GPIO.IN)
  GPIO.setup(prt_i4,GPIO.IN)
  GPIO.setup(prt_ro1,GPIO.OUT,initial=GPIO.LOW)
  GPIO.setup(prt_ro2,GPIO.OUT,initial=GPIO.LOW)
  GPIO.setup(prt_ro3,GPIO.OUT,initial=GPIO.LOW)
  GPIO.setup(prt_ro4,GPIO.OUT,initial=GPIO.LOW)
  GPIO.setup(prt_lo1,GPIO.OUT,initial=GPIO.LOW)
  GPIO.setup(prt_lo2,GPIO.OUT,initial=GPIO.LOW)
  GPIO.setup(prt_lo3,GPIO.OUT,initial=GPIO.LOW)
  GPIO.setup(prt_lo4,GPIO.OUT,initial=GPIO.LOW)
else:
  status = portio.ioperm(LPTADDRESSES[prt_lpt],1,1)
  if status:
    print("ERROR #17: Cannot access I/O port:",hex(LPTADDRESSES[prt_lpt]));
    sys.exit(17)
  status = portio.ioperm(LPTADDRESSES[prt_lpt] + 1, 1, 1)
  if status:
    print("ERROR #17: Cannot access I/O port:",hex(LPTADDRESSES[prt_lpt] + 1));
    sys.exit(17)
  portio.outb(0,LPTADDRESSES[prt_lpt])

while True:
  print(" * What do you like?")
  selection = input(" \
   1: Check I1-4 inputs\n \
   2: Check RO1-4 relay contact outputs\n \
   3: Check LO1-4 open collector outputs\n \
   4: Write text to the mini serial console\n \
   q: Quit\n")
  if selection is "Q" or selection is "q":
    print(" * Quitting.")
    if HW == 0:
      GPIO.cleanup()
    else:
      portio.outb(0,LPTADDRESSES[prt_lpt])
    sys.exit(0)

  if selection is "1":
    print(" * Check I1-4 inputs")
    if HW == 0:
      print("   used GPIO ports:")
      print("     I1: GPIO", prt_i1,sep = '')
      print("     I2: GPIO", prt_i2,sep = '')
      print("     I3: GPIO", prt_i3,sep = '')
      print("     I4: GPIO", prt_i4,sep = '')
      print("   Press ^C to stop!")
      try:
        while True:
          s = ""
          if GPIO.input(prt_i4):
            s = "1"
          else:
            s = "0"
          if GPIO.input(prt_i3):
            s = s + "1"
          else:
            s = s + "0"
          if GPIO.input(prt_i2):
            s = s + "1"
          else:
            s = s + "0"
          if GPIO.input(prt_i1):
            s = s + "1"
          else:
            s = s + "0"
          print(s)
          time.sleep(1)
      except KeyboardInterrupt:
        print()
    else:
      print("   used lines of LPT port:")
      print('     LPT #',prt_lpt + 1,sep = '')
      print("     I1: -ERROR")
      print("     I2: SELECT")
      print("     I3: PE")
      print("     I4: -ACKNOLEDGE")
      print("   Press ^C to stop!")
      try:
        while True:
          indata = portio.inb(LPTADDRESSES[prt_lpt] + 1)
          indata = indata << 1
          indata = indata >> 4
          print("    ",format(indata,'04b'))
          time.sleep(1)
      except KeyboardInterrupt:
        print()

  if selection is "2":
    print(" * Check RO1-4 relay contact outputs")
    if HW == 0:
      print("   used GPIO ports:")
      print("     RO1: GPIO", prt_ro1,sep = '')
      print("     RO2: GPIO", prt_ro2,sep = '')
      print("     RO3: GPIO", prt_ro3,sep = '')
      print("     RO4: GPIO", prt_ro4,sep = '')
      print("   Press ^C to stop!")
      try:
        while True:
          GPIO.output(prt_ro1,1)
          time.sleep(1)
          GPIO.output(prt_ro1,0)
          GPIO.output(prt_ro2,1)
          time.sleep(1)
          GPIO.output(prt_ro2,0)
          GPIO.output(prt_ro3,1)
          time.sleep(1)
          GPIO.output(prt_ro3,0)
          GPIO.output(prt_ro4,1)
          time.sleep(1)
          GPIO.output(prt_ro4,0)
      except KeyboardInterrupt:
        GPIO.output(prt_ro1,0)
        GPIO.output(prt_ro2,0)
        GPIO.output(prt_ro3,0)
        GPIO.output(prt_ro4,0)
        print()
    else:
      print("   used lines of LPT port:")
      print('     LPT #',prt_lpt + 1,sep = '')
      print("     RO1: D0")
      print("     RO2: D1")
      print("     RO3: D2")
      print("     RO4: D3")
      print("   Press ^C to stop!")
      try:
        while True:
          portio.outb(1,LPTADDRESSES[prt_lpt])
          time.sleep(1)
          portio.outb(2,LPTADDRESSES[prt_lpt])
          time.sleep(1)
          portio.outb(4,LPTADDRESSES[prt_lpt])
          time.sleep(1)
          portio.outb(8,LPTADDRESSES[prt_lpt])
          time.sleep(1)
      except KeyboardInterrupt:
          portio.outb(0,LPTADDRESSES[prt_lpt])
          print()

  if selection is "3":
    print(" * Check LO1-4 open collector outputs")
    if HW == 0:
      print("   used GPIO ports:")
      print("     LO1: GPIO", prt_lo1,sep = '')
      print("     LO2: GPIO", prt_lo2,sep = '')
      print("     LO3: GPIO", prt_lo3,sep = '')
      print("     LO4: GPIO", prt_lo4,sep = '')
      print("   Press ^C to stop!")
      try:
        while True:
          GPIO.output(prt_lo1,1)
          time.sleep(1)
          GPIO.output(prt_lo1,0)
          GPIO.output(prt_lo2,1)
          time.sleep(1)
          GPIO.output(prt_lo2,0)
          GPIO.output(prt_lo3,1)
          time.sleep(1)
          GPIO.output(prt_lo3,0)
          GPIO.output(prt_lo4,1)
          time.sleep(1)
          GPIO.output(prt_lo4,0)
      except KeyboardInterrupt:
        GPIO.output(prt_lo1,0)
        GPIO.output(prt_lo2,0)
        GPIO.output(prt_lo3,0)
        GPIO.output(prt_lo4,0)
        print()
    else:
      print("   used lines of LPT port:")
      print('     LPT #',prt_lpt + 1,sep = '')
      print("     LO1: D4")
      print("     LO2: D5")
      print("     LO3: D6")
      print("     LO4: D7")
      print("   Press ^C to stop!")
      try:
        while True:
          portio.outb(16,LPTADDRESSES[prt_lpt])
          time.sleep(1)
          portio.outb(32,LPTADDRESSES[prt_lpt])
          time.sleep(1)
          portio.outb(64,LPTADDRESSES[prt_lpt])
          time.sleep(1)
          portio.outb(128,LPTADDRESSES[prt_lpt])
          time.sleep(1)
      except KeyboardInterrupt:
          portio.outb(0,LPTADDRESSES[prt_lpt])
          print()

  if selection is "4":
    print(" * Write text to the mini serial console")
    print("   used COM port:")
    print("     device: ", prt_com)
    print("     speed:  ", com_speed)
    print("   Press ^C to stop!")
    try:
      while True:
        com.open
        com.write(str.encode(input() + EOL))
        com.close
    except KeyboardInterrupt:
        print()

sys.exit(0)
